kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/lemmas.k

```diff
@@ -48,23 +48,62 @@ module LEMMAS-WITHOUT-SLOT-UPDATES [symbolic]
-
-    rule bool2Word(A) |Int bool2Word(B) => bool2Word(A  orBool B) [simplification]
-    rule bool2Word(A) &Int bool2Word(B) => bool2Word(A andBool B) [simplification]
-
-    rule 1 |Int bool2Word(_B) => 1            [simplification]
-    rule 1 &Int bool2Word( B) => bool2Word(B) [simplification]
-
     // #newAddr range
     rule 0 <=Int #newAddr(_,_)             => true   [simplification]
     rule         #newAddr(_,_) <Int pow160 => true   [simplification]
     rule         #newAddr(_,_) <Int pow256 => true   [simplification]
 
     rule #isPrecompiledAccount(#newAddr(_, _), _) => false [simplification]
 
+  // ########################
+  // #rangeBool reasoning
+  // ########################
+
+    rule [rangeBool-not-zero-l]: notBool (X ==Int 0) => X ==Int 1 requires #rangeBool(X) [simplification]
+    rule [rangeBool-not-zero-r]: notBool (0 ==Int X) => X ==Int 1 requires #rangeBool(X) [simplification]
+    rule [rangeBool-not-one-l]:  notBool (X ==Int 1) => X ==Int 0 requires #rangeBool(X) [simplification]
+    rule [rangeBool-not-one-r]:  notBool (1 ==Int X) => X ==Int 0 requires #rangeBool(X) [simplification]
+
+  // ########################
+  // bool2Word reasoning
+  // ########################
+
+    // Range
+    rule [b2w-lb]: 0 <=Int bool2Word(_)         => true [simplification, smt-lemma]
+    rule [b2w-ub]:         bool2Word(_) <=Int 1 => true [simplification, smt-lemma]
+
+    // Relationship with equality
+    rule [b2w-eq-zero]: bool2Word(X) ==Int 0 => notBool X [simplification(30), comm]
+    rule [b2w-eq-one]:  bool2Word(X) ==Int 1 => X         [simplification(30), comm]
+
+    // Relationship with other comparators
+    rule [b2w-lt-one]:  bool2Word(B) <Int 1 => notBool B [simplification(30)]
+    rule [b2w-gt-zero]: 0 <Int bool2Word(B) => B         [simplification(30)]
+
+    rule [b2w-lt-true]:  bool2Word(_) <Int X:Int => true  requires 1 <Int X [simplification(30), concrete(X)]
+    rule [b2w-gt-false]: X:Int <Int bool2Word(_) => false requires 1 <Int X [simplification(30), concrete(X)]
+
+    // Relationship with `&Int`
+    rule [b2w-band-one]:           1  &Int bool2Word(B) => bool2Word(B)           [simplification]
+    rule [b2w-band]:     bool2Word(A) &Int bool2Word(B) => bool2Word(A andBool B) [simplification]
+
+    // Relationship with `|Int`
+    rule [b2w-bor-one]:           1  |Int bool2Word(_) => 1                     [simplification]
+    rule [b2w-bor]:     bool2Word(A) |Int bool2Word(B) => bool2Word(A orBool B) [simplification]
+
+    // Relationship with `xorInt`
+    rule [b2w-xor-zero]:           0  xorInt bool2Word(Y) => bool2Word(Y)                                                     [simplification, comm]
+    rule [b2w-xor-one]:            1  xorInt bool2Word(Y) => 1 -Int bool2Word(Y)                                              [simplification, comm]
+    rule [b2w-xor]:      bool2Word(X) xorInt bool2Word(Y) => bool2Word ( (X andBool notBool Y) orBool (notBool X andBool Y) ) [simplification]
```

Comment:
Do we avoid using `xorBool` on the RHS on purpose?

Reply:
I didn't actually know it existed - it's very rarely used. Knowing that, if some simplifications get rid of the `bool2Word` and start reasoning about the parameters, I think I trust more the reasoning we have about `andBool` and `orBool` than that we have about `xorBool`.